### ISSUE TITLE
Provenance and metadata

### ISSUE THREAD

**Issue by jeromekelleher** [tskit-dev/msprime#284]
We need some method to record provenance and metadata for tree sequences. I propose two things:

1. Add a ``ProvenanceTable`` that records where a particular tree sequence came from and what operations have been performed on it.
2. Add a 'metadata' column to (e.g.) ``NodeTable`` and ``SiteTable`` to allow us to record metadata about particular sites and nodes.

This is increasingly important now that a number of different programs can produce tree sequences. It is particularly important when storing real data, which we will be doing shortly.

## ProvenanceTable

There is already a half-hearted effort to so this in mpsrime, where we record a list of 'provenance strings' for each tree sequence. The idea here is to  allow us to store some information that would allow the specific simulation to be reproduced. An example of a provenance string for msprime is
```json
{
  "environment": {
    "python": {
      "implementation": "CPython",  "version": ["2", "7", "12" ]
    },
    "os": {
      "node": "empire",
      "release": "4.4.0-97-generic",
      "machine": "x86_64",
      "version": "#120-Ubuntu SMP Tue Sep 19 17:28:18 UTC 2017",
      "system": "Linux"
    },
    "gsl": {"version": [ 2, 1]},
    "hdf5": { "version": [ 1, 8,16]}
  },
  "version": "0.4.1.dev723+ng503704e",
  "command": "simulate",
  "parameters": {
    "TODO": "encode simulation parameters"
  },
  "software": "msprime"
}
```
The format of the JSON isn't important here, but the idea is for a particular program to encode all the necessary information to reproduce a particular simulation. 

My proposal is to create a ProvenanceTable to explicitly store these provenance strings, and to provide an easy API for adding new rows. This would have two columns:

```
timestamp                                record
2017-10-19T07:08:49+00:00   {"environment"... "command":"simulate", "parameters":{...}, "software": "ftprime"}
2017-10-19T07:08:50+00:00   {"environment"... "command":"simplify", "parameters":{"samples":[0, 1]}, software:"msprime"}
```

Here we ran a simulation in ftprime, and then ran 'simplify' on the resulting tree sequence. The idea is that every time we make a copy of a tree sequence and run some operation on it, we also copy the provenance table and add a new row describing the operation. This should provide a useful ledger of the operations that have been performed to create a particular tree sequence, and in principle, make it fully reproducible.

I think it's better to keep the format of the table as simple as possible, and push all of the real information into the metadata. It's much easier to iterate on a good JSON schema for representing this information than it is to add and remove columns from a table. The timestamp field (in ISO8601 format) is good to have I think. 

The format of the JSON encoding the metadata isn't important at this point I think. The main questions are (a) is this a good way of recording this information; and (b) is the two column (timestamp, metadata) approach good?

## Site metadata

We will want to record site-specific metadata somewhere, so adding a column to the ``SiteTable`` seems like a good way to do this. This would look something like
```
position    ancestral_state      metadata
1234        A                   {"s": 0.1, "h":0.1}
1235        C                   {"s": 100, "h":100} 
```
Here we've encoded the sites and included the parameters used to simulate them. We can also imaging putting in stuff from a VCF INFO field in here more generally.

So, basically I'm proposing that  we add an optional column to SiteTable that is a variable length string. Client code can put whatever it wants in there, but JSON seems like a good option.

**Comment by petrelharp** [comment]
The site metadata proposal sounds good; I thought about instead having another table with one column 'site' and then arbitrary other columns, but although that would be easier to extract info from, it'd be a good deal less flexible.

But: for `s` and `h` shouldn't really the metadata be in the `MutationTable`?  I guess there would be use cases for metadata columns in both.

**Comment by molpopgen** [comment]
Does the metadata have to be string-like (or dict-like as in the SiteTable example)?  Or can any valid Python object be placed there?

**Comment by jeromekelleher** [comment]
> Does the metadata have to be string-like (or dict-like as in the SiteTable example)? Or can any valid Python object be placed there?

I think storing bytes values is the best approach. That way the user can store whatever they want in there: JSON, pickled Python objects or even something like protobuf. Since we're just storing raw bytes there's no Python dependency here and can be used from any language.

**Comment by molpopgen** [comment]
> I think storing bytes values is the best approach. That way the user can store whatever they want in there: JSON, pickled Python objects or even something like protobuf. Since we're just storing raw bytes there's no Python dependency here and can be used from any language.

That makes sense.  There will likely be a small cost of data replication (storing a pickled mutation will replicate position info, etc.), but that shouldn't be a deal-breaker.

**Comment by jeromekelleher** [comment]
PR #322 adds full support for metadata on nodes, sites and mutations. Metadata is stored as a bytes object, so there's full flexibility for the user in how they encode the data. I think we can do some nice things by allowing the user to register a decoder function with the tree sequence so that they can access their metadata easily via attribute access.

What sort of stuff do you have in mind for storing as metadata @molpopgen? It would be helpful to have a few examples to work with for documentation and stuff.

**Comment by molpopgen** [comment]
> What sort of stuff do you have in mind for storing as metadata @molpopgen? It would be helpful to have a few examples to work with for documentation and stuff.

For mutations, we will need something that allows re-creation of the original type.  I'm imagining pickling the mutations directly from the C++ side of the simulation, allowing them to be restored later.

I can pickle any type in C++ whose Python representation defines __getstate__/__setstate__:
https://github.com/molpopgen/fwdpy11/blob/dev/tests/pickling_cpp.cpp#L25

Unit test here:

https://github.com/molpopgen/fwdpy11/blob/dev/tests/test_pickling.py#L34

The two links above are from the dev branch of fwdpy11.  They'll be merged into master in the new year (I hope!).  I'm working on many back-end changes to how the C++ code is organized, in order to prep the package for adding ancestry tracking in early 2018.

For nodes, the situation is less elegant.  Two nodes equals one diploid, but the meta-data are the same for the two nodes.  For ancient samples, the gamete records of a diploid become meaningless.  Thus, I plan to pickle a tuple of the diploid meta-data, which are the readonly fields of the Python type: 

https://github.com/molpopgen/fwdpy11/blob/dev/fwdpy11/src/fwdpy11_types.cc#L114

I'll probably end up adding a function to return such a tuple from an instance, and it will probably include other things like the generation in the sim (forwards in time???) when the pickling happened.  Unpickling the metadata would then restore the tuple and the user will have to recognize which node pairs belong to the same diploid.

Looking forward to trying this!  Once we submit the ftprime paper, I'll likely augment the ARG tracking example repo to attempt to use the mutation meta data.

**Comment by jeromekelleher** [comment]
Thanks @molpopgen, this is very useful. So, you're planing on pickling a bunch of stuff in there. This is good to know, and means we need to ensure binary encodings are supported from the start. (This is only really awkward when thinking about the text file format, otherwise it's just treated as a bunch of bytes)

> For nodes, the situation is less elegant. Two nodes equals one diploid, but the meta-data are the same for the two nodes.

This is a good point. I guess in the next version we could add another table ``Individual`` or something, which groups together haploid nodes. The NodeTable would get an 'individual' column then. This is definitely something we have to put off until after we've released the current state of things though --- dealing with ploidy properly isn't something to be taken lightly!

> Looking forward to trying this! Once we submit the ftprime paper, I'll likely augment the ARG tracking example repo to attempt to use the mutation meta data.

Sounds good. I was planning on making a feature of this in the paper too: recording metadata is something most bioinformatics stuff fails abysmally at. We don't have to say much, just point out that the tables API has slots available to allow recording of arbitrary metadata.

**Comment by molpopgen** [comment]
I think pickling makes the most sense.  I could do C++ strings and rely on casts to `bytes` or `str` options, but then one ends up with text data rather than objects on the user-facing side, which seems clunky.

**Comment by jeromekelleher** [comment]
This has been done.
